### PR TITLE
OPT: cmdline

### PR DESCRIPTION
**Pull request by yarikoptic**
This pull request addresses some items in #2494 and adds sugaring like

```shell
$> time datalad instal -s /// /tmp/ds2                    
Unknown command instal.
 Did you mean install or uninstall?
 All known commands: add, add-archive-content, add-readme, addurls, aggregate-metadata, annotate-paths, check-dates, clean, clone, create, create-sibling, create-sibling-github, create-test-dataset, diff, download-url, drop, export-archive, export-to-figshare, extract-metadata, get, install, ls, metadata, no-annex, publish, remove, rerun, run, save, search, siblings, sshrun, subdatasets, test, uninstall, unlock, update, wtf
datalad instal -s /// /tmp/ds2  0.59s user 0.07s system 99% cpu 0.661 total
```
(in such cases I do not see a point of spiting out regular summary on options).  We could have done similar thing for options themselves but I did not see any way to extract options from the parser

Core commands now should "fire up" and provide help in up to ~300ms (so there is room for improvement) on my laptop (instead of a second or so)

I do feel that it is "piling up"... partial parsing is also done later in `main()` itself but I didn't want to dig up there for this one.

### Changes
- [x] This change is complete

Please have a look @datalad/developers

### CODE REVIEW

**Comment by yarikoptic** [comment]
BTW:
- not sure if I've not ruined manpage generation
- not sure if even without this PR we would not be generating manpages for all extensions... which I guess we shouldn't since they should be generated/provided by the extensions packages (if I got the idea right)

**Comment by yarikoptic** [comment]
Added a hint on where to find some commands:
```shell
$> datalad crawl-init                                                                
Unknown command crawl-init.
 All known commands: add, add-archive-content, add-readme, addurls, aggregate-metadata, annotate-paths, check-dates, clean, clone, create, create-sibling, create-sibling-github, create-test-dataset, diff, download-url, drop, export-archive, export-to-figshare, extract-metadata, get, install, ls, metadata, no-annex, publish, remove, rerun, run, save, search, siblings, sshrun, subdatasets, test, uninstall, unlock, update, wtf
Hint: Command crawl-init is provided by (not installed) extension datalad-crawler, which you can install with your package manager or pip
```
A mouthful -- feel welcome to improve

**Comment by yarikoptic** [comment]
Almost there! Some interaction with ssh runner to check

**Comment by kyleam** [comment]
Here's the test that's failing on several of the builds:

```
FAIL: datalad.support.tests.test_sshrun.test_exit_code
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/virtualenv/python3.5.5/lib/python3.5/site-packages/nose/case.py", line 198, in runTest
    self.test(*self.arg)
  File "/home/travis/build/datalad/datalad/datalad/tests/utils.py", line 944, in newfunc
    return func(*args, **kwargs)
  File "/home/travis/build/datalad/datalad/datalad/support/tests/test_sshrun.py", line 34, in test_exit_code
    assert_equal(cme.exception.code, 42)
AssertionError: 1 != 42
```

File: datalad/cmdline/main.py [hunk 1]
```diff
@@ -193,49 +191,96 @@ def setup_parser(
...
+            from ..interface import _known_extension_commands
+            extension_commands = {
+                c: e
+                for e, commands in _known_extension_commands.items()
+                for c in commands
+            }
+            hint = None
+            if unparsed_arg in extension_commands:
+                hint = "Command %s is provided by (not installed) extension %s," \
+                       " which you can install with your package manager or pip" \
```
Comment (kyleam) [inline datalad/cmdline/main.py]: IMO the ", which ..." doesn't add much information and could be dropped.

File: datalad/cmdline/main.py [hunk 2]
```diff
@@ -277,18 +319,63 @@ def setup_parser(
...
+        # just to appear in print_usage also consistent with --help output
+        parser.add_argument("command [command-opts]")
+        parser.print_usage(file=sys.stderr)
+    else:
+        sys.stderr.write("Unknown %s %s.\n" % (what, provided,))
+        import difflib
+        suggestions = difflib.get_close_matches(provided, known)
+        if suggestions:
+            sys.stderr.write(" Did you mean %s?\n" % " or ".join(suggestions))
+        sys.stderr.write(" All known %ss: %s\n" % (what, ", ".join(sorted(known))))
```
Comment (kyleam) [inline datalad/cmdline/main.py]: I wonder if we should bother filling the screen up with all known commands.

Here's what git does:

```
% git dff
git: 'dff' is not a git command. See 'git --help'.

The most similar command is
        diff
```

And git-annex

```
% git annex addulr
git-annex: Unknown command 'addulr'

Did you mean one of these?
        addurl
        adjust

```

Comment (yarikoptic) [reply datalad/cmdline/main.py]: ok, married the two (but didn't use annex'es "dump all commands" if no suggestion found):
```shell
$> datalad lkjsldkfjsdf
datalad: Unknown command 'lkjsldkfjsdf'.  See 'datalad --help'.

$> datalad installl    
datalad: Unknown command 'installl'.  See 'datalad --help'.

Did you mean one of these?
        install
        uninstall
```
and also will use "known" extensions in suggestions:
```shell
$> datalad crawll
datalad: Unknown command 'crawll'.  See 'datalad --help'.
 
Did you mean one of these?
        crawl
        crawl-init

$> datalad crawl   
datalad: Unknown command 'crawl'.  See 'datalad --help'.

Hint: Command crawl is provided by (not installed) extension datalad-crawler.
```

Will add a basic test and push...

**Comment by mih** [comment]
Manpages still look good -- or at least not worse ;-)

**Comment by mih** [comment]
Shorter help with command suggestions feels good!

File: datalad/interface/__init__.py [hunk 3]
```diff
@@ -81,3 +81,9 @@
         ('datalad.support.sshrun', 'SSHRun', 'sshrun'),
         ('datalad.distribution.subdatasets', 'Subdatasets', 'subdatasets'),
     ])
+
+# Some known extensions and their commands to suggest whenever lookup fails
+_known_extension_commands = {
+    'datalad-crawler': ('crawl', 'crawl-init'),
+    'datalad-neuroimaging': ('bids2scidata',)
+}
```
[the remark this answers is not in the document]
Comment (yarikoptic) [reply datalad/interface/__init__.py]: `crawl` commands is a mere regression from previous versions -- someone who upgrades and if they used crawl before would be surprised to not find it.
false positives/negatives -- I am not expecting any within a year or two

**Comment by mih** [comment]
So be it.

**Pull request merged**